src/app/icd/IcdMonitoringTable.cpp

```diff
@@ -99,37 +188,36 @@ CHIP_ERROR IcdMonitoringTable::Set(uint16_t index, const IcdMonitoringEntry & en
     VerifyOrReturnError(index < this->Limit(), CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrReturnError(kUndefinedNodeId != entry.checkInNodeID, CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrReturnError(kUndefinedNodeId != entry.monitoredSubject, CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrReturnError(entry.key.size() == IcdMonitoringEntry::kKeyMaxSize, CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrReturnError(nullptr != entry.key.data(), CHIP_ERROR_INVALID_ARGUMENT);
     IcdMonitoringEntry e(this->mFabric, index);
     e.checkInNodeID    = entry.checkInNodeID;
     e.monitoredSubject = entry.monitoredSubject;
-    e.key              = entry.key;
     e.index            = index;
+    memcpy(e.key.AsMutable<Crypto::Aes128KeyByteArray>(), entry.key.As<Crypto::Aes128KeyByteArray>(),
+           Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
+
     return e.Save(this->mStorage);
 }
 
 CHIP_ERROR IcdMonitoringTable::Remove(uint16_t index)
 {
-    IcdMonitoringEntry entry(this->mFabric, index);
+    IcdMonitoringEntry entry(mSessionKeyStore, this->mFabric, index);
 
     // Shift remaining entries down one position
     while (CHIP_NO_ERROR == this->Get(static_cast<uint16_t>(index + 1), entry))
     {
-        // WARNING: The key data is held in the entry's serializing buffer
-        IcdMonitoringEntry copy = entry;
-        this->Set(index++, copy);
+        this->Set(index++, entry);
     }
 
     // Remove last entry
     entry.fabricIndex = this->mFabric;
     entry.index       = index;
+    ReturnErrorOnFailure(entry.DeleteKey());
```

Comment:
This is worth some comments.  Isn't that the same key as the previous index is using?  Or did we end up getting a new handle as part of the Set() above?  Should document.

Reply:
fixed and documented